drivers/mcp2515/mcp2515_spi.c

```diff
@@ -0,0 +1,163 @@
+
+    /* Configure SPI */
+    res = spi_init_cs(dev->conf->spi, dev->conf->cs_pin);
+    if (res != SPI_OK) {
+        DEBUG("spi_init_master: error initializing SPI_%i device (code %i)\n",
+              dev->conf->spi, res);
+        return -1;
+    }
+    return 0;
+}
+
+int mcp2515_spi_reset(const candev_mcp2515_t *dev)
+{
+    spi_acquire(dev->conf->spi, dev->conf->cs_pin, dev->conf->spi_mode,
+                dev->conf->spi_clk);
+    spi_transfer_byte(dev->conf->spi, dev->conf->cs_pin, false,
+                      MCP2515_SPI_RESET);
+    spi_release(dev->conf->spi);
+    return 0;
+}
+
+int mcp2515_spi_read(const candev_mcp2515_t *dev, uint8_t addr, uint8_t *buf,
+                     unsigned int len)
+{
+    spi_acquire(dev->conf->spi, dev->conf->cs_pin, dev->conf->spi_mode,
+                dev->conf->spi_clk);
+    spi_transfer_byte(dev->conf->spi, dev->conf->cs_pin, true,
+                      MCP2515_SPI_READ);
+    spi_transfer_regs(dev->conf->spi, dev->conf->cs_pin, addr, NULL,
+                      (void *)buf, len);
+    spi_release(dev->conf->spi);
+    return 0;
+}
+
+int mcp2515_spi_read_rxbuf(const candev_mcp2515_t *dev, uint8_t mailbox,
+                           void *buf, uint8_t len)
+{
+    spi_acquire(dev->conf->spi, dev->conf->cs_pin, dev->conf->spi_mode,
+                dev->conf->spi_clk);
+    spi_transfer_byte(dev->conf->spi, dev->conf->cs_pin, true,
+                      MCP2515_SPI_READ_RXBUF | (mailbox << 2));
+    spi_transfer_bytes(dev->conf->spi, dev->conf->cs_pin, false, NULL,
+                       (void *)buf, len);
+    spi_release(dev->conf->spi);
+    return 0;
+}
+
+int mcp2515_spi_write(const candev_mcp2515_t *dev, uint8_t addr, uint8_t *buf,
+                      unsigned int len)
+{
+    spi_acquire(dev->conf->spi, dev->conf->cs_pin, dev->conf->spi_mode,
+                dev->conf->spi_clk);
+    spi_transfer_byte(dev->conf->spi, dev->conf->cs_pin, true,
+                      MCP2515_SPI_WRITE);
+    spi_transfer_regs(dev->conf->spi, dev->conf->cs_pin, addr, (void *)buf,
+                      NULL, len);
+    spi_release(dev->conf->spi);
+    return 0;
+}
+
+int mcp2515_spi_write_txbuf(const candev_mcp2515_t *dev, uint8_t mailbox,
+                            void *buf, uint8_t len)
+{
+    spi_acquire(dev->conf->spi, dev->conf->cs_pin, dev->conf->spi_mode,
+                dev->conf->spi_clk);
+    spi_transfer_byte(dev->conf->spi, dev->conf->cs_pin, true,
+                      MCP2515_SPI_LOAD_TXBUF | (mailbox << 1));
+    spi_transfer_bytes(dev->conf->spi, dev->conf->cs_pin, false, (void *)buf,
+                       NULL, len);
+    spi_release(dev->conf->spi);
+    return 0;
+}
+
+int mcp2515_spi_rts(const candev_mcp2515_t *dev, uint8_t mailbox)
+{
+    spi_acquire(dev->conf->spi, dev->conf->cs_pin, dev->conf->spi_mode,
+                dev->conf->spi_clk);
+    spi_transfer_byte(dev->conf->spi, dev->conf->cs_pin, false,
+                      MCP2515_SPI_RTS | (1 << mailbox));
```

Comment:
Why the mailbox shift?

Reply:
From the datasheet:
> The RTS  command  can  be  used  to  initiate  message transmission for one or more of the transmit buffers. The MCP2515 is selected by lowering the CS pin. The RTS   command   byte   is   then   sent.   As   shown   in Figure 12-6,  the  last  3  bits  of  this  command  indicate which transmit buffer(s) are enabled to send.

Shifting with mailbox selects which mailbox is ready to receive.